op-service/sources/l1_beacon_client.go

```diff
@@ -0,0 +1,157 @@
+package sources
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/crypto/kzg4844"
+
+	"github.com/ethereum-optimism/optimism/op-service/client"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+)
+
+type L1BeaconClient struct {
+	cl client.HTTP
+
+	initLock     sync.Mutex
+	timeToSlotFn TimeToSlotFn
+}
+
+// NewL1BeaconClient returns a client for making requests to an L1 consensus layer node.
+func NewL1BeaconClient(cl client.HTTP) *L1BeaconClient {
+	return &L1BeaconClient{cl: cl}
+}
+
+func (cl *L1BeaconClient) apiReq(ctx context.Context, dest any, method string) error {
+	headers := http.Header{}
+	headers.Add("Accept", "application/json")
+	resp, err := cl.cl.Get(ctx, method, headers)
+	if err != nil {
+		return fmt.Errorf("%w: http Get failed", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		errMsg, _ := io.ReadAll(resp.Body)
+		_ = resp.Body.Close()
+		return fmt.Errorf("failed request with status %d: %s", resp.StatusCode, string(errMsg))
+	}
+	if err := json.NewDecoder(resp.Body).Decode(dest); err != nil {
+		_ = resp.Body.Close()
+		return err
+	}
+	if err := resp.Body.Close(); err != nil {
+		return fmt.Errorf("%w: failed to close response body", err)
+	}
+	return nil
+}
+
+type TimeToSlotFn func(timestamp uint64) uint64
+
+// GetTimeToSlotFn returns a function that converts a timestamp to a slot number.
+func (cl *L1BeaconClient) GetTimeToSlotFn(ctx context.Context) (TimeToSlotFn, error) {
+	cl.initLock.Lock()
+	defer cl.initLock.Unlock()
+	if cl.timeToSlotFn != nil {
+		return cl.timeToSlotFn, nil
+	}
+
+	var genesisResp eth.APIGenesisResponse
+	if err := cl.apiReq(ctx, &genesisResp, "eth/v1/beacon/genesis"); err != nil {
+		return nil, err
+	}
+
+	var configResp eth.APIConfigResponse
+	if err := cl.apiReq(ctx, &configResp, "eth/v1/config/spec"); err != nil {
+		return nil, err
+	}
+
+	cl.timeToSlotFn = func(timestamp uint64) uint64 {
+		return (timestamp - uint64(genesisResp.Data.GenesisTime)) / uint64(configResp.Data.SecondsPerSlot)
+	}
+	return cl.timeToSlotFn, nil
+}
+
+// GetBlobSidecars fetches blob sidecars that were confirmed in the specified L1 block with the
+// given indexed hashes. Order of the returned sidecars is not guaranteed, and blob data is not
+// checked for validity.
+func (cl *L1BeaconClient) GetBlobSidecars(ctx context.Context, ref eth.L1BlockRef, hashes []eth.IndexedBlobHash) ([]*eth.BlobSidecar, error) {
+	if len(hashes) == 0 {
+		return []*eth.BlobSidecar{}, nil
+	}
+	slotFn, err := cl.GetTimeToSlotFn(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to get time to slot function", err)
+	}
+	slot := slotFn(ref.Time)
+
+	builder := strings.Builder{}
```

Comment:
Seems more canonical to use the url package here?
- https://pkg.go.dev/net/url#URL.JoinPath
- https://pkg.go.dev/net/url#Values

Reply:
I was thinking something like

```go
url.Join(sidecarsMethod, strconv.FormatUint(slot, 10))
```

but not super opinionated if its not something you want to do

Reply:
chatgpt suggested the following code

```go
package main

import (
	"fmt"
	"net/url"
)

func main() {
	// Step 1: Create the base URL
	baseURL, err := url.Parse("https://example.com")
	if err != nil {
		panic(err)
	}

	// Step 2: Prepare query parameters
	params := url.Values{}
	params.Add("param1", "value1")
	params.Add("param2", "value2")
	params.Add("param3", "value3")

	// Step 3: Encode and attach the query parameters
	baseURL.RawQuery = params.Encode()

	// The final URL with query parameters
	fmt.Println(baseURL.String())
}
```

Reply:
It may error in `url.Parse` if it doesn't have the `http` prefix so not that big of a deal, this is consensus critical code so i am just being nit picky

Reply:
Unlike the value encoder, I'm not seeing much improvement in code clarity from using JoinPaths here TBH, and the additional robustness it provides isn't really needed as all fragments are known safe to just blindly append. YMMY of course.  I did realize I forgot to pull out this particular method prefix constant, so have fixed that.